driver-core/src/main/com/mongodb/internal/connection/InternalStreamConnection.java

```diff
@@ -319,23 +319,18 @@ public boolean isClosed() {
     @Override
     public <T> T sendAndReceive(final CommandMessage message, final Decoder<T> decoder, final SessionContext sessionContext,
             final RequestContext requestContext) {
-        ByteBufferBsonOutput bsonOutput = new ByteBufferBsonOutput(this);
         CommandEventSender commandEventSender;
 
-        try {
+        try (ByteBufferBsonOutput bsonOutput = new ByteBufferBsonOutput(this)) {
```

Comment:
I'm surprised that this works.  `bsonOutput` is passed to `sendCommandMessage`, which takes responsibility for calling `close`.  So aren't we double-closing here?  It doesn't look like `ByteBufferBsonOutput` protects against that.

Reply:
There's a good reason for this, too.  If the message needs to be compressed, we want to release the buffers holding the uncompressed message once we have compressed it into a new set of buffers.

Reply:
> `bsonOutput` is passed to `sendCommandMessage`, which takes responsibility for calling close

Right, one of the examples where non-local knowledge is required to reason about correctness. I paid attention to this on my own, and fortunately, `ByteBufferBsonOutput.close` must be idempotent because `Closeable.close` is specified as such: _"If the stream is already closed then invoking this method has no effect."_ (the Java SE API is clearly documented incorrectly by implying that any `Closeable` implementation is a stream, but the idempotence is nonetheless there). Indeed, `ByteBufferBsonOutput.close` clears its `bufferList` to avoid releasing buffers more than once.

Reply:
OK, I totally overlooked the buffer list clearing when I glanced at the close method.  I was expecting something like `if (!closed)`.

Still I wonder if we should modify this method in the same commit.  Is this fixing an actual bug, or clarifying intent?  If the latter, maybe we should just document the intent about who is responsible for closing the instance.

Reply:
Since I don't think that trying to handle `Error`s is reasonable (an we don't do that in many cases), this change does not fix a bug: try-finally is not better than try-catch(RuntimeException) in this case. So it only clarifies the intent. However, the method `send(final CommandMessage message, final Decoder<T> decoder, final SessionContext sessionContext)` located right after this one also uses try-with-resources block to close `bsonOutput` and also calls `sendCommandMessage`. The new code not only makes the intent cleaner, it makes the code consistent with that method.

If at this point you still want me to roll the changes in this method back, I will.

Reply:
I would prefer it, yes.  I think we should open a separate PR to refactor the code to clarify intent.

Reply:
Done. The changes are not in https://github.com/mongodb/mongo-java-driver/pull/884.